### PR TITLE
Rejoin and privacy

### PR DESCRIPTION
**Pull request by dromedarCase**
The bot will now unban users and send them an invite link, when requested, after a specified time. Also in order to prevent leakage of invitation links to this repository `constants.js` was added to `.gitignore` and moved to `constants.js.default`.

### CODE REVIEW

File: constants.js.default [hunk 1]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+    answer: 'Würden Sie es mir übel nehmen, sollte ich Sie darum bitten Ihre Fresse zu halten?',
+    banTime: 10, // seconds
```
Comment (rkr8) [inline constants.js.default]: `banTime` should probably be renamed to `banTimeInSeconds`.

File: constants.js.default [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+    invitation: 'Hier könnte Ihre Werbung stehen.', //TODO: make invitation dependent on group being banned from
```
Comment (rkr8) [inline constants.js.default]: `invitationText` would be more appropriate.

File: constants.js.default [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module.exports = {
+    answer: 'Würden Sie es mir übel nehmen, sollte ich Sie darum bitten Ihre Fresse zu halten?',
+    banTime: 10, // seconds
+    invitation: 'Hier könnte Ihre Werbung stehen.', // TODO: make invitation dependent on group being banned from
```
Comment (rkr8) [inline constants.js.default]: `invitationText` would be more appropriate.

**Comment by rkr8** [comment]
Could you please explain, which invitation links you mean?

File: messageLog.js [hunk 4]
```diff
@@ -1,12 +1,17 @@
 module.exports = function (size) {
     return {
+        banned: [],
```
Comment (rkr8) [inline messageLog.js]: `banned` should get a seperate module.

File: messageLog.js [hunk 5]
```diff
@@ -1,12 +1,17 @@
...
         index: 0,
         messages: new Array(size),
         addMessage: function (msg) {
             // minimal ring buffer implementation
             this.messages[this.index] = msg;
             this.index = (this.index + 1) % this.logSize;
+        },
+        addBanned: function (id) {
+            var date = new Date()/1000;
+            this.banned.push([id, date]);
```
Comment (rkr8) [inline messageLog.js]: Maybe should use something like this:
```
this.banned.push(
 {
  userId: id,
  date: new Date() / 1000
 }
);
```

File: registerActions.js [hunk 6]
```diff
@@ -2,6 +2,21 @@ module.exports = function () {
     this.on('message', function (msg) {
         // filter images, stickers, etc.
         // TODO: prevent sticker spamming
+        
+        // send unbanned user invite link
+        this.messageLog.banned.forEach(
+            function (item) {
```
Comment (rkr8) [inline registerActions.js]: `item` should be replaced with `user`.

File: registerActions.js [hunk 7]
```diff
@@ -2,6 +2,21 @@ module.exports = function () {
...
+                if (item[0] == msg.from.id && new Date()/1000-item[1] > this.bot.constants.banTime) {
```
Comment (rkr8) [inline registerActions.js]: A comment would make this more readable.

File: registerActions.js [hunk 8]
```diff
@@ -2,6 +2,21 @@ module.exports = function () {
...
+                    this.bot.invite (msg);
+                    item[0] = null; // TODO: find a cleaner, yet easy, way to remove user from list
+                }
+            }.bind(
```
Comment (rkr8) [inline registerActions.js]: A comment should explain why `bind` is needed.

File: constants.js.default (hunk 1 above)
Comment (dromedarCase) [inline constants.js.default]: I think the shorter name would make the constant more usable in the future.

File: constants.js.default (hunk 3 above)
Comment (dromedarCase) [inline constants.js.default]: Right.

File: messageLog.js (hunk 4 above)
Comment (dromedarCase) [inline messageLog.js]: Maybe, but I think it might be more convenient to keep the list of banned users in the same module the messages are logged in, since both will have to have an instance for each chat the bot is in, eventually.

File: messageLog.js (hunk 5 above)
Comment (dromedarCase) [inline messageLog.js]: Yes, but it's essentially the same. I think this should be done whenever we will add more parameters.

File: registerActions.js (hunk 6 above)
Comment (dromedarCase) [inline registerActions.js]: No, it shouldn't. It's not the users we are iterating over. It's tupels containing information about the message, which triggered the ban.

File: registerActions.js (hunk 7 above)
Comment (dromedarCase) [inline registerActions.js]: ...or not. It's kind of obvious.

File: registerActions.js (hunk 8 above)
Comment (dromedarCase) [inline registerActions.js]: rtfm

File: answerAndKick.js [hunk 9]
```diff
@@ -16,7 +16,10 @@ module.exports = function answerAndKick(msg, answer) {
     }, function (error) {
         // user is probably admin
         console.log(chalk.red(logSymbols.error, error));
+        return;
```
Comment (dromedarCase) [inline answerAndKick.js]: This just ends the execution of the function declared in line 16.

File: answerAndKick.js [hunk 10]
```diff
@@ -16,7 +16,10 @@ module.exports = function answerAndKick(msg, answer) {
...
     });
+    
+    // user was successfully kicked
```
Comment (dromedarCase) [inline answerAndKick.js]: No, they weren't.

File: constants.js.default (hunk 1 above)
Comment (rkr8) [inline constants.js.default]: In contrast to `banTime`, `banTimeInSeconds` is self-explanatory. The shorter name is probably more usable for you, but not for new contributors.

File: messageLog.js (hunk 4 above)
Comment (rkr8) [inline messageLog.js]: One module should fulfill exactly one task. `messageLog.js` is responsible for storing messages in a ring buffer. Hence the name.

File: messageLog.js (hunk 5 above)
Comment (rkr8) [inline messageLog.js]: I think, multidimensional lists should be used in other cases.

File: registerActions.js (hunk 6 above)
Comment (rkr8) [inline registerActions.js]: *tuples
No, `item` is not a tuple. Tuples are fixed in size. Besides that, `item` is a very vague identifier.

**Comment by dromedarCase** [comment]
Done. Although I don't see how this was a blocker.

File: registerActions.js (hunk 7 above)
Comment (rkr8) [inline registerActions.js]: future-proof. Beside that, those comments form the basis for the documentation.

File: registerActions.js (hunk 8 above)
Comment (rkr8) [inline registerActions.js]: Same as above. Think of the documentation.

File: constants.js.default (hunk 1 above)
Comment (dromedarCase) [inline constants.js.default]: According to your reasoning, `answer` should be renamed to `messageToBeSendToUserBeforeTheyGetBannedSoTheyKnowItWasABot`

File: messageLog.js (hunk 4 above)
Comment (dromedarCase) [inline messageLog.js]: Alright. I'll fix it, when I will extend it to differentiate between groups.

File: messageLog.js (hunk 5 above)
Comment (dromedarCase) [inline messageLog.js]: That's the only good use for a multidimensional list.

File: answerAndKick.js (hunk 9 above)
Comment (rkr8) [inline answerAndKick.js]: The problem is that an `error` does not stop the ban process. Therefore, invitation links are also sent to users who have not been banned successfully. This could lead to a memory leak.

File: registerActions.js (hunk 6 above)
Comment (dromedarCase) [inline registerActions.js]: When iterating through any kind of items, calling these items 'items' is a perfectly reasonable thing.

PS:
Beginning your comment with an asterisk is being a dick to your correspondent.

File: registerActions.js (hunk 7 above)
Comment (dromedarCase) [inline registerActions.js]: I don't see how you would comment on `item[0] == msg.from.id && new Date()/1000-item[1] > this.bot.constants.banTime`, while not just rephrasing it in plain English.

File: constants.js.default (hunk 1 above)
Comment (rkr8) [inline constants.js.default]: Nope, I said the identifier should contain the unit so that users don't even need to read the comment.

File: messageLog.js (hunk 5 above)
Comment (rkr8) [inline messageLog.js]: Please read [this](http://www.geekinterview.com/question_details/20396) or any other article about multidimensional arrays/lists.

File: registerActions.js (hunk 6 above)
Comment (rkr8) [inline registerActions.js]: As you said above, we iterate through “(...)tupels containing information about the message, which triggered the ban(...)” and not through “any kind of items”.

File: constants.js.default (hunk 1 above)
Comment (dromedarCase) [inline constants.js.default]: Same as with `answer`. You should not have to read the description, if you can't read the comment.

File: registerActions.js (hunk 8 above)
Comment (dromedarCase) [inline registerActions.js]: rtf**m**

File: answerAndKick.js (hunk 9 above)
Comment (dromedarCase) [inline answerAndKick.js]: Right. But this doesn't help it.

File: messageLog.js (hunk 5 above)
Comment (dromedarCase) [inline messageLog.js]: Have _you_ read it? This _exactly_ where you would use one.

File: registerActions.js (hunk 6 above)
Comment (dromedarCase) [inline registerActions.js]: Tell me how these tuples aren't items.

**Comment by dromedarCase** [comment]
Screw it. It isn't leading anywhere.

**Comment by rkr8** [comment]
In my opinion, it isn't a good idea to store invitation links in a secret file in order to prevent a “(...)leakage of invitation(...)”. That would mean, that the bot uses the same invitation link for different users.

**Comment by rkr8** [comment]
By the way, that's why I asked “Could you please explain, which invitation links you mean?”.

**Comment by rkr8** [comment]
My other comments were just general topics, which weren't related to the change request.
